### PR TITLE
Avoid overflow in vector_norm for scalar input

### PR DESCRIPTION
**Pull request by isuruf**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #144073

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

Fixes https://github.com/pytorch/pytorch/issues/143960 where torch.dist gave different results from eager due to vector_norm overflowing and eager mode avoids the overflow for single element reductions by not computing the power and then the root.

### CODE REVIEW

File: test/inductor/test_torchinductor.py [hunk 1]
```diff
@@ -2510,6 +2510,17 @@ def fn(a, b):
 
         self.common(fn, (torch.randn(4, 4), torch.randn(4, 4)))
 
+    @skip_if_halide  # different pow accuracies
```
Comment (amjames) [inline test/inductor/test_torchinductor.py:2513]: is the accuracy delta large enough that we don't support it? Or should this conditionally modify the tolerance if the halide back-end is selected?

File: torch/_refs/linalg/__init__.py [hunk 2]
```diff
@@ -151,6 +151,23 @@ def vector_norm(
         reduce_sum = partial(torch.sum, dim=dim, keepdim=keepdim)
 
         is_ord_even = ord % 2 == 0 if isinstance(ord, IntLike) else ord % 2.0 == 0.0
+        if (dim is None and guard_size_oblivious(x.numel() == 1)) or (
+            dim is not None
+            and (x.ndim > 0 and all(guard_size_oblivious(x.shape[d] == 1) for d in dim))
+        ):
+            x = torch.abs(x)
+            if x.ndim > 64:
```
Comment (amjames) [inline torch/_refs/linalg/__init__.py:159]: NIT: We should apply this check before we take the absolute value. No sense in doing the work if we are failing anyway.

File: test/inductor/test_torchinductor.py (hunk 1 above)
Comment (isuruf) [reply test/inductor/test_torchinductor.py:2513]: It's a big difference.
```
Mismatched elements: 3 / 16 (18.8%)
Greatest absolute difference: 0.06096780672669411 at index (1, 0) (up to 1e-05 allowed)
Greatest relative difference: 1.1311241388320923 at index (1, 0) (up to 1.3e-06 allowed)
```

**Review by eellison — COMMENTED** [review]
Would you mind linking me to the eager logic that does this ?

**Review by eellison — APPROVED** [review]

File: torch/_refs/linalg/__init__.py [hunk 3]
```diff
@@ -151,6 +151,23 @@ def vector_norm(
         reduce_sum = partial(torch.sum, dim=dim, keepdim=keepdim)
 
         is_ord_even = ord % 2 == 0 if isinstance(ord, IntLike) else ord % 2.0 == 0.0
+        if (dim is None and guard_size_oblivious(x.numel() == 1)) or (
```
Comment (laithsakka) [inline torch/_refs/linalg/__init__.py:154]: we are moving away from guard_size_oblivious, a better solution here is to put this check x.numel() == 1
in the torch.check() so that in case its unbacked we do not throw data dependent error.

**Review by laithsakka — CHANGES_REQUESTED** [review]

File: torch/_refs/linalg/__init__.py (hunk 3 above)
Comment (laithsakka) [reply torch/_refs/linalg/__init__.py:154]: see 
https://github.com/pytorch/pytorch/pull/148809
and for more context
this doc
https://docs.google.com/presentation/d/1kxhT96nn3y053aOmqXfKuMR-KD1g0Xcvl_gL3ePs-Lc/edit?usp=sharing

Comment (eellison) [reply torch/_refs/linalg/__init__.py:154]: In this case @laithsakka, we need this for numerics.

Comment (eellison) [reply torch/_refs/linalg/__init__.py:154]: We can't avoid guards if it means that the numerics will be off.

Comment (laithsakka) [reply torch/_refs/linalg/__init__.py:154]: same apply to x.shape[d] == 1 checks

Comment (isuruf) [reply torch/_refs/linalg/__init__.py:154]: Yes, I think a plain `x.numel() == 1` is better.

**Review by laithsakka — APPROVED** [review]
